src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,325 @@
+            nodes = nodes[numNodes:]
+            instancesTerminate = [instance for instance, nodeInfo in nodes]
+        else:
+            # Without load info all we can do is sort instances by time left in billing cycle.
+            instances = sorted(instances,
+                               key=lambda instance: (CGCloudProvisioner._remainingBillingInterval(instance)))
+            instancesTerminate = [instance for instance in islice(instances, numNodes)]
+        if instancesTerminate:
+            self._deleteIAMProfiles(instances=instancesTerminate, ctx=self.ctx)
+            self._terminateInstance(instances=instancesTerminate, ctx=self.ctx)
+        else:
+            logger.debug('No nodes to delete')
+        return len(instancesTerminate)
+
+    @staticmethod
+    def __getNodesInCluster(ctx, securityGroupName, preemptable=False, both=False):
+        pendingInstances = ctx.ec2.get_only_instances(filters={'instance.group-name': securityGroupName,
+                                                               'instance-state-name': 'pending'})
+        runningInstances = ctx.ec2.get_only_instances(filters={'instance.group-name': securityGroupName,
+                                                               'instance-state-name': 'running'})
+        instances = set(pendingInstances)
+        if not preemptable and not both:
+            return [x for x in instances.union(set(runningInstances)) if x.spot_instance_request_id is None]
+        elif preemptable and not both:
+            return [x for x in instances.union(set(runningInstances)) if x.spot_instance_request_id is not None]
+        elif both:
+            return [x for x in instances.union(set(runningInstances))]
+
+    def _getNodesInCluster(self, preeptable=False, both=False):
+        if not both:
+            return self.__getNodesInCluster(self.ctx, self.securityGroupName, preemptable=preeptable)
+        else:
+            return self.__getNodesInCluster(self.ctx, self.securityGroupName, both=both)
+
+    def _getWorkersInCluster(self, preemptable):
+        entireCluster = self._getNodesInCluster(both=True)
+        logger.debug('All nodes in cluster %s', entireCluster)
+        workerInstances = [i for i in entireCluster if i.private_ip_address != self.masterIP and
+                           preemptable != (i.spot_instance_request_id is None)]
+        logger.debug('Workers found in cluster after filtering %s', workerInstances)
+        return workerInstances
+
+    @staticmethod
+    def _getSpotRequestIDs(ctx, securityGroupName):
+        requests = ctx.ec2.get_all_spot_instance_requests()
+        return [x.id for x in requests if x.launch_group == securityGroupName]
+
+    @staticmethod
+    def _createSecurityGroup(ctx, name):
+        # security group create/get. ssh + all ports open within the group
+        web = ctx.ec2.create_security_group(name, 'Toil appliance security group')
```

Comment:
I've had one hell of a time with setting up a security group when it already existed.